### DISCUSSION TITLE
Automation and Integration of Structural Analysis and Design Processes Using BHoM

### DISCUSSION THREAD

**Discussion by fatimaaboubane** [BHoM/BHoM#1602] [Q&A] [answered]
Hello everyone, as PhD students focused on structural engineering, we are exploring methods to automate and integrate structural analysis and design processes using the Building Information Modelling Hierarchy of Needs (BHoM). Our inquiries include:

Is there a method or script available to automatically generate load combinations in Robot Structural Analysis (RSA) using BHoM parameters, simplifying the workflow for structural analysis and design?
How can wind and snow loads be automatically generated for steel frame buildings within the BHoM framework, leveraging scripting or parametric modeling techniques?
What are the best practices for incorporating local codes, such as BAEL (French building code), into BHoM enumerations and workflows, ensuring compliance and accuracy in structural analysis and design?
We are seeking insights, tools, resources, and best practices from the GitHub community to address these challenges and advance our research on automated structural verifications and dimensioning.

Tags: BHoM, Structural Engineering, Automation, Robot Structural Analysis, Wind Loads, Snow Loads, BAEL, Local Codes, Scripting, Parametric Modeling

**Comment by FraserGreenroyd** [comment, marked as answer]
Hi @fatimaaboubane , it's great to hear the BHoM is being given some research space - I hope we'll be able to provide some useful feedback to support you! 😄 

These questions are best answered by our structural development community members which I'll tag here and also ping directly to make sure they see it!

@IsakNaslundBh @peterjamesnugent @al-fisher @rwemay @pawelbaran

**Reply by fatimaaboubane** [reply]
Hi, @FraserGreenroyd thank you very much for your response and encouragement, it should be an honor to hear from the Bhom experts.

**Comment by peterjamesnugent** [comment]
Hi @fatimaaboubane, great to hear you are looking in to the BHoM for use in your research!

### Load Combinations
You can [push](https://bhom.xyz/documentation/BHoM_Adapter/Adapter-Actions/#an-example-the-push-action) `LoadCombination`s to Robot and read them back if needed.

See here for a full list of objects that are currently supported by the Robot _Toolkit: https://github.com/BHoM/Robot_Toolkit/wiki/BHoM-Robot-object-relations#loading

Whilst the common usert interface (UI) is Grasshopper, Excel is also supported. For something like `Loadcases`s and `LoadCombination`s, Excel is better suited (in my experience) because load combinations are thought of in a tabular way.

There is some guidance on the general principles to get started with [Visual Programming](https://bhom.xyz/documentation/Guides-and-Tutorials/Visual-Programming-with-BHoM/).

A workflow might look like:
1. On one sheet, have the [properties of a `Loadcase`](https://github.com/BHoM/BHoM/blob/develop/Structure_oM/Loads/Loadcase.cs) represented by columns, and create a new `Loadcase` on each row;
2. On a second sheet, transpose the `Loadcase`s and have `LoadCombination`s represented by each row;
3. Then it becomes a case of just filling in the factors for each `LoadCombination`;
4. You can then push the `LoadCombination`s (and the Robot_Adapter) will push the `Loadcase`s too because they are a depedency type:
https://github.com/BHoM/Robot_Toolkit/blob/744f0f7d6fbb96d77b5a9953e023e19b12c1f762/Robot_Adapter/Types/DependencyTypes.cs#L51

You can read more about [depedency types](https://bhom.xyz/documentation/BHoM_Adapter/Implement-an-Adapter/#dependency-types).

If you had a standard set of load combinations (for example from a code practice) that you wanted to use across mulitple models, one option would be to store the `LoadCombination` as a JSON file.

Any `BHoMObject` can be serialised as a JSON file, in fact all of our datasets are stored in this way. For example the [UK Steel Sections](https://github.com/BHoM/BHoM_Datasets/tree/develop/DataSets/Structure/SectionProperties/UK_SteelSections). 

So you could determine `LoadCombination` once, and store it as a JSON for everyone in your project to then use. There's a bit of information about [creation of datasets](https://bhom.xyz/documentation/Guides-and-Tutorials/Coding-with-BHoM/Datasets/Creation-of-Datasets/) that you may find useful.

### Snow Loading and Wind Loading
If you are modelling [`Panel`](https://github.com/BHoM/BHoM/blob/develop/Structure_oM/Elements/Panel.cs)s then you will need to use an [`AreaUniformlyDistributedLoad`](https://github.com/BHoM/BHoM/blob/develop/Structure_oM/Loads/AreaUniformlyDistributedLoad.cs).

If you are using only [`Bar`](https://github.com/BHoM/BHoM/blob/develop/Structure_oM/Elements/Bar.cs)s then you will need to use [`BarUniformlyDistributedLoad`](https://github.com/BHoM/BHoM/blob/develop/Structure_oM/Loads/BarUniformlyDistributedLoad.cs) for the snow load and [`BarVaryingDistributedLoad`](https://github.com/BHoM/BHoM/blob/develop/Structure_oM/Loads/BarVaryingDistributedLoad.cs) for the wind loads (if they are not uniformly distributed).

These can both be setup parametrically and tailored for each project or model. If you know the pressure you want to apply, you can parametrically calculate the UDL based on the `SectionProperty` of the `Bar`.

### Enumeration and incorperating local codes 
If there are particular tables, or values you want to store - one option would be to use a dataset. Or you could store all the values as a JSON file that others can read from.

### Examples and best practice
We have a [samples](https://github.com/BHoM/samples) repositry that is worth checking out. This is a recent [end to end workflow using Lusas but will work just as well for Robot.](https://github.com/BHoM/samples/blob/main/Lusas_Toolkit/BHoM_ExampleScript_PushExecutePull.gh).

I have linked quite a few specific pages from the [documentation](https://bhom.xyz/documentation/), but it worth going through as it details the principle concepts, how the BHoM was concieved and how to get involved.

If you find anything missing that you want to add, please see how to [edit the documentation](https://bhom.xyz/documentation/Contributing/Editing-the-documentation/) or [raise an issue](https://github.com/BHoM/documentation/issues).

Let me know if there's anything I have missed or you wish to discuss further.

**Reply by fatimaaboubane** [reply]
Hi @peterjamesnugent 
![combin](https://github.com/BHoM/BHoM/assets/159909096/867c2970-178c-4663-b072-c9ede06e509b)
![pull](https://github.com/BHoM/BHoM/assets/159909096/f67723e3-3f4e-4e45-b770-ce686d2e8d2c)
![Q](https://github.com/BHoM/BHoM/assets/159909096/349fa26b-9e4a-4d18-8ea9-290841a2bdba)


I hope this message finds you well. I wanted to extend my sincere gratitude for your invaluable assistance in guiding me through the development of my BHoM script. Your instructions have proven to be incredibly helpful, and I greatly appreciate the time and effort you've dedicated to helping me navigate through the complexities of the script.

I'm writing to you with some additional queries and challenges that I've encountered while working on my script. I hope you can provide further insights or guidance to help me address these issues effectively.

Combination Formatting in Robot: I've noticed that when I manually add combinations such as DEP and EFF in Robot for my steel frame building, they appear in one format (ElU only). However, I intend for these combinations to be represented as more distinct categories, including DEP-, DEP, DEP+, EFF+, EFF-, and EFF. Could you please advise on how to achieve this parametrically using BHoM?

Wind Load Representation: Another issue I've encountered is with wind load representation. While Robot lists cases as load cases, in BHoM, manually input values seem to represent just one application case. How can I ensure that wind load representations in BHoM align more closely with the progressive categorizations typically calculated in Robot?

Exploration Load Application: Additionally, I've been trying to apply exploration load to panels within my script. However, when attempting to pull them, I've encountered difficulties as it doesn't seem to work as intended. Could you provide any insights or recommendations on how to address this issue effectively?

I truly appreciate your ongoing support and expertise in navigating these technical challenges. Your insights have been invaluable to my learning journey, and I'm eager to continue refining my BHoM script with your guidance.

Thank you once again for your assistance and support. I look forward to hearing from you soon and continuing our collaboration.

Warm regards,

**Reply by peterjamesnugent** [reply]
Hi @fatimaaboubane, not a problem. I am glad the wiki and examples are proving useful.

### LoadCombination
There is not a parameter for the CombinationType defined on the `LoadCombination` at the moment but I agree it would be useful to have, not least to determine whether a combination is a combonation or an envelope. Actually looks like there was some discussion on this: https://github.com/BHoM/BHoM/issues/1195 already. I will plan this in for the next Sprint (starting 25th March 2024) - it would need to be implemented for each structural adapter. Would you be interested adding the functionality for Robot_Toolkit?

When pushing it would be a change here:
https://github.com/BHoM/Robot_Toolkit/blob/744f0f7d6fbb96d77b5a9953e023e19b12c1f762/Robot_Adapter/CRUD/Create/Loads/LoadCombination.cs#L46

And pulling it would be here:
https://github.com/BHoM/Robot_Toolkit/blob/develop/Robot_Adapter/CRUD/Read/Loads/LoadCombinations.cs

### Loads
I am not sure I fully understand the question here. You can assign multiple loads i.e. `AreaUniformlyDistributedLoad` and `BarDistributedLoad` to a single `Loadcase`. You can also assign multiple objects to a load as long as the `Vector` is the same. Lastly, you should make sure each load has a unique name. If I have missed the ball completely, please share an example model/script.

I have not come across exploration loads before, are they similar to loads applied using a search area? If you have an example, I can take a quick look.

Thanks,

Peter

**Reply by fatimaaboubane** [reply]
hello, Mr @peterjamesnugent I hope you are well.
For the type of combination it will be very beneficial and very interesting for me to integrate them into Robot Toolkit; as identifying the exact nature of the combination will define the weightings and simplify the sizing; on the other hand, the operating forces are linked to the type of building and its use as a preliminary solution it can be defined as UDL and linked to structural panels as a slab; for the value we can define load values linked to each type of building in the dictionary function.
For wind and snow loads we have always been able to find a method to automate the generation of the wind model, without doing it manually on a robot.
Thank you very much for your interest, it is an honor to exchange in this issue.

**Comment by dextertuay99** [comment]
Hi @fatimaaboubane, have you discovered any way to remove/replace the Loadcase? Cause I found out that when I push the Loadcase to RSA, it tends to add another BarDistributedLoad to my structure.

**Reply by fatimaaboubane** [reply]
Hi @dextertuay99 , the only situation where my script adds another bar distributed load is when the push of loads depends on the success of the model push operation, in this case the solution could be to add independent push components for the modulation and loading phases. If this is not the case, feel free to write to me with further explanations of the problem you are encountering. I would like to find a solution and help you get on with your BHoM work.
